nfs/src/rpc_task.cpp

```diff
@@ -911,23 +913,47 @@ static void createfile_callback(
             UPDATE_INODE_ATTR(inode, res->CREATE3res_u.resok.dir_wcc.after);
         }
 
-        /*
-         * TODO: Remove this assert and if server doesn't send the filehandle,
-         *       make a LOOKUP RPC request to query the filehandle and pass
-         *       that to fuse. Since this callback is called in the context of
-         *       the libnfs thread, we should not issue a sync LOOKUP.
-         *       Same needs to be done for other create APIs like mkdir,
-         *       symlink and mknod.
-         */
-        assert(
-            res->CREATE3res_u.resok.obj.handle_follows &&
-            res->CREATE3res_u.resok.obj_attributes.attributes_follow);
+        if (!res->CREATE3res_u.resok.obj.handle_follows) {
+            /*
+             * If the server doesn't send the filehandle (which is perfectly
+             * valid), make a LOOKUP RPC request to query the filehandle and
+             * pass that to fuse.
+             */
+            AZLogWarn("CreateFile failed to return filehandle, req={}, "
+                "parent_ino={}, name={}. Issuing lookup.",
+                fmt::ptr(task->get_fuse_req()),
+                task->rpc_api->create_task.get_parent_ino(),
+                task->rpc_api->create_task.get_file_name());
 
-        task->get_client()->reply_entry(
-            task,
-            &res->CREATE3res_u.resok.obj.post_op_fh3_u.handle,
-            &res->CREATE3res_u.resok.obj_attributes.post_op_attr_u.attributes,
-            task->rpc_api->create_task.get_fuse_file());
+            struct rpc_task *lookup_tsk =
+                    task->get_client()->get_rpc_task_helper()->alloc_rpc_task(
+                        FUSE_LOOKUP);
+
+            lookup_tsk->init_lookup(
+                task->get_fuse_req(),
+                task->rpc_api->create_task.get_file_name(),
+                task->rpc_api->create_task.get_parent_ino());
+
+            // Mark the lookup task as being called for the create file.
+            lookup_tsk->rpc_api->lookup_task.set_is_called_for_create_file(
+                task->rpc_api->create_task.get_fuse_file());
+
+            lookup_tsk->run_lookup();
+
+            /*
```

Comment:
Free the current RPC task before trying to create a new lookup task as we may enter into a deadlock where the lookup task can never be created if we have no free slots and the current task can never be freed as we are blocked waited for a slot to be available.